File: leetcode/Construct_Binary_Tree_from_Preorder_and_Inorder_Traversal.java
import java.util.*;
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    int[] preorder;
    int[] inorder;
    HashMap<Integer, Integer> map = new HashMap<Integer, Integer>();
    
    TreeNode buildTree(int pa, int pb, int ia, int ib) {
        if (pa > pb || ia > ib) return null;
        TreeNode r = new TreeNode(preorder[pa]);
        int m = map.get(r.val);
        int l = m - ia;
        r.left = buildTree(pa + 1, pa + l, ia, m - 1);
        r.right = buildTree(pa + l + 1, pb, m + 1, ib);
        return r;
    }
    
    public TreeNode buildTree(int[] __preorder, int[] __inorder) {
        preorder = __preorder;
        inorder = __inorder;
        for (int i = 0; i < inorder.length; ++i) {
            map.put(inorder[i], i);
        }
        return buildTree(0, preorder.length - 1, 0, inorder.length - 1);
    }
}

File: leetcode/Construct_Binary_Tree_from_Preorder_and_Inorder_Traversal.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
    TreeNode *build(vector<int> &preorder, int begin_pre, int end_pre,
                    vector<int> &inorder,  int begin_in,  int end_in) {
        // Assume the input is valid.
        if (begin_pre >= end_pre)
            return NULL;
        vector<int>::iterator iter_mid = find(inorder.begin(),
                                              inorder.end(),
                                              preorder[begin_pre]);
        int mid = iter_mid - inorder.begin();
        TreeNode *node = new TreeNode(inorder[mid]);
        node->left  = 
            build(preorder, begin_pre + 1, begin_pre + 1 + (mid - begin_in),
                  inorder,  begin_in,      mid);
        node->right = 
            build(preorder, begin_pre + 1 + (mid - begin_in), end_pre,
                  inorder,  mid + 1,                          end_in);
        return node;
    }
public:
    TreeNode *buildTree(vector<int> &preorder, vector<int> &inorder) {
        return build(preorder, 0, preorder.size(),
                     inorder,  0, inorder.size());
    }
};
